src/persistent_state.py

```diff
@@ -0,0 +1,168 @@
+
+        _bot_state = value
+
+    # Now write the modified `bot_state` back to disk.
+    with open(bot_state_file, "w") as f:
+        # We add indenting (which also adds newlines) into the file to make it easy for a human
+        # to look through in case of the need for debugging (the performance cost in minimal).
+        bot_state_str = json.dumps(_bot_state, indent=2)
+        f.write(bot_state_str)
+
+
+def get_state(path: Iterable[str]) -> JSON_DATA_TYPE:
+    """Retrieve persistent state at a given path.
+
+    Args:
+        path: The path to retrieve the state of.
+
+    Returns:
+        The value at the given path. None if the path does not exist, or if None was
+        literally stored at this path.
+    """
+    # If path is empty, just return the entire bot state
+    if not path:
+        return copy.deepcopy(_bot_state)
+
+    # We don't want to modify the given path directly, so we make a copy here.
+    path = list(path)
+
+    # Start at the top of the dict.
+    current_path = _bot_state
+
+    # Iterate over each item in the path until we reach the second to last item. The last item will
+    # be used as the field name in the JSON dict.
+    key = path.pop()
+    for pathname in path:
+        current_path = current_path.get(pathname)
+
+        if current_path is None:
+            # Oops, we hit a dead end.
+            return None
+
+    # Return the value under the field at the end of the given path.
+    value = current_path.get(key)
+    if value is None:
+        return None
+
+    # We explicitly return a copy of the value, otherwise any manipulations the calling function
+    # does to the value may result in the state dict being changed.
+    return copy.deepcopy(value)
+
+
+def delete_state(path: Iterable[str]) -> bool:
+    """
+    Deletes the state at the given path. Removes all saved state under the given path
+    This function writes to disk.
+
+    Args:
+        path: The path to delete the state for.
+
+    Returns:
+        True if the path was deleted, False if the path did not exist.
+    """
+    if not path:
+        # We cannot delete without a path. Deleting all state this way is not supported.
+        return False
+
+    # We don't want to modify the given path directly, so we make a copy here.
+    path = list(path)
+
+    # Get the state at this path
+    field_to_delete = path.pop()
+
+    # this should be a dict as we're one level up
+    state_at_path = get_state(path)
+    if not isinstance(state_at_path, dict):
+        # The path is invalid, as removing the last item from a path should
+        # result in a path that leads to a dict.
+        return False
+
+    if state_at_path is None:
+        # This path does not exist
+        return False
+
+    if field_to_delete not in state_at_path:
+        # This field does not exist at this path.
+        return False
+
+    del state_at_path[field_to_delete]
+
+    # Store the updated state to disk
+    set_state(path, state_at_path)
```

Comment:
Actually, shouldn't we save to disk after deleting the path in the following lines?

Reply:
This is what I initially tried, but note that `state_at_path` here is a copy of the bot state dict, not the dict itself. If we don't update the actual bot state here, then the call to `delete_state` just below will end up calling `get_state` again, yielding a dictionary without any deleted state.

I can add a comment explaining this as it is subtle.